fred2/OperatorComboBox.cpp

```diff
@@ -0,0 +1,431 @@
+{
+	CComboBox::PreSubclassWindow();
+	EnableToolTips(TRUE);
+}
+
+INT_PTR OperatorComboBox::OnToolHitTest(CPoint point, TOOLINFO *pTI) const
+{
+	// not linked yet
+	if (m_listbox.GetSafeHwnd() == NULL)
+		return -1;
+
+	CPoint screenpoint = point;
+	ClientToScreen(&screenpoint);
+
+	int item = LBItemFromPt(m_listbox.GetSafeHwnd(), screenpoint, FALSE);
+
+	if (item >= 0 && item < GetCount())
+	{
+		int op_index = (int)GetItemData(item);
+		int op_const = Operators[op_index].value;
+		auto helptext = m_help_callback(op_const);
+		if (helptext == nullptr && m_listbox.IsItemEnabled(item))	// a disabled item will show a tooltip regardless
+			return -1;
+
+		CRect rect;
+		m_listbox.GetItemRect(item, &rect);
+
+		pTI->hwnd = m_listbox.GetSafeHwnd();
+		pTI->uId = item;
+		pTI->lpszText = LPSTR_TEXTCALLBACK;
+		pTI->rect = rect;
+		return pTI->uId;
+	}
+
+	return -1;
+}
+
+OperatorComboBoxList::OperatorComboBoxList(const char* (*help_callback)(int), int opf_type)
+{
+	m_help_callback = help_callback;
+	SetOpfType(opf_type);
+}
+
+int OperatorComboBoxList::GetOpfType() const
+{
+	return m_opf_type;
+}
+
+void OperatorComboBoxList::SetOpfType(int opf_type)
+{
+	m_opf_type = opf_type;
+}
+
+//here we supply the text for the item
+BOOL OperatorComboBoxList::OnToolTipText(UINT id, NMHDR* pNMHDR, LRESULT* pResult)
+{
+	// need to handle both ANSI and UNICODE versions of the message
+	TOOLTIPTEXTA* pTTTA = (TOOLTIPTEXTA*)pNMHDR;
+	TOOLTIPTEXTW* pTTTW = (TOOLTIPTEXTW*)pNMHDR;
+
+	CString strTipText;
+	UINT_PTR nID = pNMHDR->idFrom;
+
+	// Do not process the message from built in tooltip
+	if (nID == (UINT_PTR)m_hWnd &&
+		((pNMHDR->code == TTN_NEEDTEXTA && pTTTA->uFlags & TTF_IDISHWND) ||
+		(pNMHDR->code == TTN_NEEDTEXTW && pTTTW->uFlags & TTF_IDISHWND)))
+		return FALSE;
+
+	// Get the mouse position
+	auto pMessage = GetCurrentMessage(); // get mouse pos
+	ASSERT(pMessage);
+	auto pt = pMessage->pt;
+
+	int item = LBItemFromPt(GetSafeHwnd(), pt, FALSE);
+
+	if (item >= 0 && item < GetCount())
+	{
+		int op_index = (int)GetItemData(item);
+		int op_const = Operators[op_index].value;
+		auto helptext = m_help_callback(op_const);
+
+		if ((helptext != nullptr && *helptext != '\0') || !IsItemEnabled(item))
+		{
+			SCP_string buffer;
+			if (!IsItemEnabled(item))
+			{
+				int opr_type;
+				map_opf_to_opr(m_opf_type, opr_type);
+
+				buffer = "The operator \"";
+				buffer += op_index >= 0 ? Operators[op_index].text : "<invalid operator>";
+				buffer += "\" cannot be selected because it has an incompatible return type.\r\n\tReturns: ";
+				buffer += opr_type_name(query_operator_return_type(op_const));
+				buffer += "\r\n\tExpected: ";
+				buffer += opr_type_name(opr_type);
+			}
+			else
+			{
+				buffer = helptext;	// never mind the MSVC warning; helptext will be non-null and non-empty here
+			}
+			replace_all(buffer, "\t", "    ");
+			m_tooltiptextA = buffer.c_str();
+			m_tooltiptextW = buffer.c_str();
+
+#ifndef _UNICODE
```

Comment:
Since the code in both ifndef parts looks identical, could we not just leave it?

Reply:
Hmm, looks like you're right.